Open PDF attachments in the viewer instead of an unconditional download

Rob--W opened the pull request:
To test:

1. Open the PDF from #6643.
2. Click on the attachment list.
3. Click on a PDF, and confirm that it opens in a new tab.

Tested with Firefox 51 (generic target), Chrome 56 (extension).

Fixes #6643
Relevant for #4282

Rob--W committed:
Recognize file name in reference fragment in getPDFFileNameFromURL

The regular expression incorrectly marked a group as capturing.
For `http://example.com/#file.pdf`, the expected result is "file.pdf",
but instead "document.pdf" was returned.

Rob--W committed:
[CRX] Recognize blob and data-URLs in the router

When a blob or data-URL is opened with the extension, viewer.html
rewrites the URL. But when the viewer is refreshed (e.g. F5), Chrome
would fail to display the viewer because the extension router was not
set up to recognize such URLs.

Now it is.

Snuffleupagus commented on web/pdf_attachment_viewer.js:
```diff
@@ -89,6 +89,34 @@ var PDFAttachmentViewer = (function PDFAttachmentViewerClosure() {
+      var blobUrl;
+      button.onclick = function() {
+        if (!blobUrl) {
+          blobUrl = pdfjsLib.createObjectURL(content, 'application/pdf') +
```
Nit: Similar to e.g. https://github.com/mozilla/pdf.js/blob/a719b71e5957ee513ebf43a4ff703926c629b553/web/download_manager.js#L83-L84, you may want to add `pdfjsLib.PDFJS.disableCreateObjectURL` as the third parameter here!?

Rob--W replied:
Done.

Snuffleupagus commented on web/pdf_attachment_viewer.js:
```diff
@@ -89,6 +89,34 @@ var PDFAttachmentViewer = (function PDFAttachmentViewerClosure() {
+            '?file=' + encodeURIComponent(blobUrl);
+        } else {
+          // Let Firefox's content handler catch the URL and display the PDF.
+          viewerUrl = blobUrl;
```
In the Firefox addon, the `#` used above causes the viewer to try and goto a named destination equal to the `filename`. Most likely such a destination won't exist, but it'd be great to try and avoid this altogether.

Rob--W replied:
Oh I see. It turns out that there are two ways to obtain the file name.

- `getFilenameFromUrl` (originally introduced in 83b6eaed1602ec3ddcbfe6633b943ecfb3da3e10)
- `getPDFFileNameFromURL` (84ae29c21d0bcdad09cb8ff71d53d4a149100fac from #3161)

I have updated the PR to use the latter.

timvandermeij reviewed (commented):
This patch already looks good, but I would like to see the code with the existing comments addressed before approving this change. Thanks!

Rob--W committed:
Detect download filename based on full URL

Rob--W committed:
Set title using logic similar as download name

The download method (and the PDF document properties) detect the
file name using `getPDFFileNameFromURL`. The title ought to also
display the PDF filename if available.

Rob--W committed:
Open PDF attachments in the viewer instead of download

If users want to download, they can quickly click on the Download button
in the newly opened viewer.

The blobUrl logic for Firefox relies on `disableCreateObjectURL` is
never false in Firefox. If the assumption is invalid, then PDF
attachments at the attachment view will not correctly be displayed,
because a data-URL will be generated and `?<filename>` is treated as
part of the data:-URL.

Snuffleupagus commented on web/pdf_attachment_viewer.js:
```diff
@@ -89,6 +89,43 @@ var PDFAttachmentViewer = (function PDFAttachmentViewerClosure() {
+          // eslint-disable-next-line no-undef
+          viewerUrl = chrome.runtime.getURL('/content/web/viewer.html') +
+            '?file=' + encodeURIComponent(blobUrl + '#' + filename);
+        } else {
```
Do we actually need to use such complicated logic below, since this code block really only ought to be relevant for `FIREFOX`/`MOZCENTRAL` builds, where `URL.createObjectURL` should always be supported anyway?

I.e. wouldn't it be sufficient to do something like the below instead, and just ignore the `pdfjsLib.PDFJS.disableCreateObjectURL === true` case; given that there's no way for a user of the addon/built-in version to disable `createObjectURL` short of editing the source code.
```javascript
} else if (PDFJSDev.test('FIREFOX || MOZCENTRAL')) {
  // Let Firefox's content handler catch the URL and display the PDF.
  viewerUrl = blobUrl + '?' + encodeURIComponent(filename);
}
```

Rob--W replied:
I'm fine with that suggestion. Is `PDFJS.disableCreateObjectURL` mainly intended for use in websites?

Snuffleupagus replied:
> Is PDFJS.disableCreateObjectURL mainly intended for use in websites?

It's my understanding that its only reason for existence is for compatibility purposes (with IE, refer to PR #4103), see https://github.com/mozilla/pdf.js/blob/master/web/compatibility.js#L458-L465.

Rob--W replied:
I added a separate commit that removes the `disableCreateObjectURL` check.

Snuffleupagus replied:
> I added a separate commit that removes the disableCreateObjectURL check.

To me, it doesn't look entirely necessary for this to be a separate commit. Hence I'd suggest just folding into the previous commit, which also help slightly with https://github.com/mozilla/pdf.js/pull/8024#issuecomment-278182396.

Rob--W replied:
Done. I initially kept them separate in case someone ever has a need for disabling `createObjectURL` in Firefox (and wants to be able to revert the commit), but now I have squashed the two commits with an extra comment and updated commit message.

timvandermeij commented:
Could you squash the commits a bit? Having six commits seems a bit much for this relatively small change. Let's try to reduce it to two or three commits if possible.

Rob--W commented:
Each individual commit serves their own purpose, so I'd like to keep them separate. If the commit descriptions are not clear I am willing to elaborate more.

timvandermeij commented:
/botio-linux preview

timvandermeij commented:
Thank you for the patch!
